src/rgw/rgw_sal_motr.cc

```diff
@@ -2192,29 +2195,63 @@ int MotrObject::read_mobj(const DoutPrefixProvider* dpp, int64_t off, int64_t en
 
 int MotrObject::get_bucket_dir_ent(const DoutPrefixProvider *dpp, rgw_bucket_dir_entry& ent)
 {
-  int rc = 0;
+  int rc = 0, ret_code = -ENOENT;
   string tenant_bkt_name = get_bucket_name(this->get_bucket()->get_tenant(), this->get_bucket()->get_name());
   string bucket_index_iname = "motr.rgw.bucket.index." + tenant_bkt_name;
-  int max = 1000;
+  int max = 2;
   vector<string> keys(max);
   vector<bufferlist> vals(max);
   bufferlist bl;
   bufferlist::const_iterator iter;
 
-  if (this->get_bucket()->get_info().versioning_status() == BUCKET_VERSIONED ||
-      this->get_bucket()->get_info().versioning_status() == BUCKET_SUSPENDED) {
+  // Check entry in the cache
+  if (this->store->get_obj_meta_cache()->get(dpp, this->get_name(), bl) == 0) {
+      iter = bl.cbegin();
+      ent.decode(iter);
+      rc = 0;
+      goto out;
+  }
+  if(this->have_instance())
+  {
+    if(this->get_key().have_null_instance())
+      {
+        // Read the null index entry
+        std::string null_version_key = this->get_name() + "/null";
+        bufferlist bl_null;
+        bufferlist::const_iterator iter1;
+        rgw_bucket_dir_entry null_key;
+        ret_code = this->store->do_idx_op_by_name(bucket_index_iname,
+                                    M0_IC_GET, null_version_key, bl_null);
+        ldpp_dout(dpp, 20) <<__func__<< ": GET null index entry, rc = "<< ret_code << dendl;
+        if (ret_code < 0)
+        {
+         ldpp_dout(dpp, 0) << __func__ << "ERROR: Key does not exist. " << rc << dendl;
+         return ret_code;
+        }
+        if(ret_code == 0)
+        {
+          bufferlist& blr = bl_null;
+          iter1 = blr.cbegin();
+          null_key.decode(iter1);
+        }
 
-    // Check entry in the cache
-    if (this->store->get_obj_meta_cache()->get(dpp, this->get_name(), bl) == 0) {
-        iter = bl.cbegin();
+        // Get content of the key and replace it's instance with null
+        // then add into cache
+        std::string prev_null_obj_key = null_key.key.name + '[' + null_key.key.instance + ']';
+        rc = this->store->do_idx_op_by_name(bucket_index_iname,
+                    M0_IC_GET, prev_null_obj_key, bl);
+        if (rc < 0)
+        {
+         ldpp_dout(dpp, 0) << __func__ << "ERROR: Key does not exist. " << rc << dendl;
+         return rc;
+        }
+        bufferlist& blr = bl;
+        iter = blr.cbegin();
         ent.decode(iter);
-        rc = 0;
+        // Put into the cache
+        this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
```

Comment:
We have to be very care full of what we are putting into the cache, I mean should it be put using the index_key (with version id, using to_str()) or just the object key. 
We should go through all the put & get cache call and decide whether to use index key or just object key. 
Otherwise we may introduce some bugs.
For this specific case we should use to_str() as the call was made to fetch the null version.
@cdeshmukh @bkirunge7 @sachinpunadikar , please provide your inputs here.